cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -736,35 +736,102 @@ private module ImplCommon {
   }
 
+  /**
+   * A node from which flow can return to the caller. This is either a regular
+   * `ReturnNode` or a `PostUpdateNode` corresponding to the value of a parameter.
+   */
+  class ReturnNodeExt extends Node {
+    ReturnNodeExt() {
+      this instanceof ReturnNode or
+      parameterValueFlowsToUpdate(_, this)
+    }
+
+    /** Gets the kind of this returned value. */
+    ReturnKindExt getKind() {
+      result = TValueReturn(this.(ReturnNode).getKind())
+      or
+      exists(ParameterNode p, int pos |
+        parameterValueFlowsToUpdate(p, this) and
+        p.isParameterOf(_, pos) and
+        result = TParamUpdate(pos)
+      )
+    }
+  }
+
+  private newtype TReturnKindExt =
+    TValueReturn(ReturnKind kind) or
+    TParamUpdate(int pos) {
+      exists(ParameterNode p | parameterValueFlowsToUpdate(p, _) and p.isParameterOf(_, pos))
+    }
+
+  /**
+   * An extended return kind. A return kind describes how data can be returned
+   * from a callable. This can either be through a returned value or an updated
+   * parameter.
+   */
+  abstract class ReturnKindExt extends TReturnKindExt {
+    /** Gets a textual representation of this return kind. */
+    abstract string toString();
+  }
+
+  class ValueReturnKind extends ReturnKindExt, TValueReturn {
+    private ReturnKind kind;
+
+    ValueReturnKind() { this = TValueReturn(kind) }
+
+    ReturnKind getKind() { result = kind }
+
+    override string toString() { result = kind.toString() }
+  }
+
+  class ParamUpdateReturnKind extends ReturnKindExt, TParamUpdate {
+    private int pos;
+
+    ParamUpdateReturnKind() { this = TParamUpdate(pos) }
+
+    int getPosition() { result = pos }
+
+    override string toString() { result = "param update " + pos }
+  }
+
   /** A callable tagged with a relevant return kind. */
   class ReturnPosition extends TReturnPosition0 {
     private DataFlowCallable c;
-    private ReturnKind kind;
+    private ReturnKindExt kind;
 
     ReturnPosition() { this = TReturnPosition0(c, kind) }
 
     /** Gets the callable. */
     DataFlowCallable getCallable() { result = c }
 
     /** Gets the return kind. */
-    ReturnKind getKind() { result = kind }
+    ReturnKindExt getKind() { result = kind }
 
     /** Gets a textual representation of this return position. */
     string toString() { result = "[" + kind + "] " + c }
   }
 
   pragma[noinline]
-  DataFlowCallable returnNodeGetEnclosingCallable(ReturnNode ret) {
+  DataFlowCallable returnNodeGetEnclosingCallable(ReturnNodeExt ret) {
     result = ret.getEnclosingCallable()
   }
 
   pragma[noinline]
-  ReturnPosition getReturnPosition(ReturnNode ret) {
-    exists(DataFlowCallable c, ReturnKind k | returnPosition(ret, c, k) |
+  ReturnPosition getReturnPosition(ReturnNodeExt ret) {
+    exists(DataFlowCallable c, ReturnKindExt k | returnPosition(ret, c, k) |
       result = TReturnPosition0(c, k)
     )
   }
 
+  Node getAnOutNodeExt(DataFlowCall call, ReturnKindExt kind) {
```

Comment:
Make this an abstract predicate defined by virtual dispatch  on `ReturnKindExt`?

Reply:
Hmm, I guess it doesn't really matter much, so sure I could do that.  I guess it should then simply be named `getAnOutNode` as the `Ext` part becomes implicit?

Reply:
👍

Reply:
Done.  Checking one dataflow query, the DIL was essentially unchanged, so I don't expect any perf surprises from this.